### ISSUE TITLE
TensorFlow 1.15 requires AVX support, which is incompatible with M1 Macs

### ISSUE THREAD

**Issue by joshuacwnewton** [spinalcordtoolbox/spinalcordtoolbox#3566]
In #3361, we upgraded Tensorflow from 1.5 to 1.15. This was necessary to move away from Python 3.6, but doing so introduced a new requirement: Users' CPUs must support the AVX instruction set. 

At the time this was acceptable to us, because the vast majority of CPUs produced after 2011 support AVX. However, M1 Macs [do not support AVX](https://www.anaconda.com/blog/apple-silicon-transition):

> For the last 10 years, Intel and AMD have rolled out support for “vector instructions,” which operate on multiple pieces of data simultaneously. Specifically, these additions to the x86-64 instruction set are AVX, AVX2, and AVX-512 (which itself has different variants). As you might imagine, these instructions can be handy when working with array data, and several libraries have adopted them when compiling binaries for x86-64. **The problem is that Rosetta2 does not support any of the AVX family of instructions and will produce an illegal instruction error if your binary tries to use them.** Because of the existence of different Intel CPUs with different AVX capabilities, many programs can already dynamically select between AVX, AVX2, AVX-512, and non-AVX code paths, detecting the CPU capabilities at runtime. These programs will work just fine under Rosetta2 because the CPU capabilities reported to processes running under Rosetta2 do not include AVX. **However, suppose an application or library does not have the capability to pick non-AVX code paths at runtime.** In that case, it will not work under Rosetta2 if the packager assumed the CPU would have AVX support when they compiled the library. **TensorFlow wheels, for example, do not work under Rosetta2 on the M1.**

This means that, unless we revert back to Tensorflow v1.5, SCT's `sct_deepseg_<x>` functions will not be compatible with M1 Macs.

* Related: #3046

**Comment by joshuacwnewton** [comment]
From @jcohenadad on Slack:

> should we consider reverting to the previous TF version? sct_deepseg is a pretty central function in SCT

This would mean reverting #3361 and going back to Python 3.6 for the release of SCT v5.5 (the version to be used in the upcoming SCT Course).

* **Pros:** We guarantee that `sct_deepseg_sc` and `sct_deepseg_gm` work for everyone who participates in the course, regardless of their hardware.
* **Cons:** We would be sticking with a version of Python that is rapidly approaching EOL. 
    * NB1: We can treat this as temporary, and we can always go back to Python 3.7 (i.e. after v5.5 and the SCT Course).
    * NB2: In truth, using a version of Python that has reached EOL is not immediately a concern, as SCT ships with its own isolated conda environment that users are likely not going to interact with beyond the CLI entry points. 

In this case, I think the pros outweigh the cons.

**Comment by joshuacwnewton** [comment]
> This would mean reverting #3361 and going back to Python 3.6 for the release of SCT v5.5 (the version to be used in the upcoming SCT Course).

Good news; v5.5 is _not_ needed for the SCT Course. v5.4 will be sufficient because of https://neuropoly.slack.com/archives/CAW9X21D5/p1636154453068000. And, v5.4 does not suffer from this issue, because the changes from #3361 are only present on master, and not in our stable release.

**Comment by kousu** [comment]
T______________________________________________________T

Why is this even an issue? M1 isn't even an intel chipset? Like that article says

> Those packages all need to be recompiled on macOS for ARM64 CPUs to run natively on the new M1-based Macs.

so is this really an issue?

Oh I see because

> Rosetta2 is that it is an x86-64 to ARM64 translator, not an emulator

So at the moment, tensorflow is getting installed thanks to an emulation (emulation, translation, close enough) layer. Emulation is always going to have compatibility bugs -- ever tried to play an old video game, or use an old Windows app via their built in old-windows emulator? Building to this spec is a house of cards.

Can we afford to just say SCT isn't compatible with M1 macs until tensorflow gets its act together? I know they're the new shiny product that everyone wants, but we have other platforms probably more widely distributed (say, WSL! e.g. #3564 and the problem (yet undocumented) I had helping Maxime this week) to worry about supporting too.

Perhaps, even, tensorflow(/numpy/pytorch) has already got their act together in the latest release, and the onus is on us to unpin from the old version?

> That said, I'm worried about the [normalization of deviance](https://en.wikipedia.org/wiki/Normalization_of_deviance), so I don't want to grow too comfortable with this solution.

Yes. Excellently phrased.

> Good news; v5.5 is _not_ needed for the SCT Course, because we will no longer need to submit an urgent release to fix https://neuropoly.slack.com/archives/CAW9X21D5/p1636154453068000.
> 
> So, we should be OK for M1 users when the SCT Course is conducted.

If so, can we skip #3568?

**Comment by kousu** [comment]
We could also solve this issue the *same way* we solved the AVX one: compile tensorflow ourselves on an M1 mac, and insert installing it into

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/88a72e07ad7593c0f26b9e04a962201b4c5a5533/install_sct#L652-L657

**Comment by jcohenadad** [comment]
> Can we afford to just say SCT isn't compatible with M1 macs until tensorflow gets its act together?

unfortunately we cannot. Among the collaborators I know who are using SCT at least 10 of them already go the new M1. So the amount of M1 users is not negligible.

**Comment by joshuacwnewton** [comment]
> > Good news; v5.5 is _not_ needed for the SCT Course, because we will no longer need to submit an urgent release to fix https://neuropoly.slack.com/archives/CAW9X21D5/p1636154453068000.
> > 
> > So, we should be OK for M1 users when the SCT Course is conducted.
> 
> If so, can we skip #3568?

We can skip #3568 only for as long as we're able stick with v5.4. 

If we were to create a new stable release from `master` in its current state, the release would be incompatible with M1 Macs. So, we would need to find some sort of solution for this issue prior to creating a new release

### Solution 1: Compiling TF for AVX-less Macs

> We could also solve this issue the same way we solved the AVX one: compile tensorflow ourselves on an M1 mac, and insert installing it into

For this suggestion, just to provide some context, the way I compiled Tensorflow was by targeting a specific older microarchtecture (bazel's [`-march`](https://github.com/spinalcordtoolbox/docker-tensorflow-builder/blob/83dee3d41ef0131b533d4a84d0843b75f6424b77/tensorflow/ubuntu-18.04/build.sh#L82) `core2`) in using an Ubuntu 18.04 docker image from [this repo](https://github.com/spinalcordtoolbox/docker-tensorflow-builder). The `bazel` environment settings I used are [here](https://github.com/spinalcordtoolbox/docker-tensorflow-builder/blob/83dee3d41ef0131b533d4a84d0843b75f6424b77/tensorflow/ubuntu-18.04/docker-compose.yml#L10):

```yaml
    environment:
      - TF_VERSION_GIT_TAG=${TF_VERSION_GIT_TAG:-v1.15.5}
      - PYTHON_VERSION=${PYTHON_VERSION:-3.7}
      - BAZEL_VERSION=${BAZEL_VERSION:-0.26.1}
      - USE_GPU=${USE_GPU:-0}
      - CUDA_VERSION=${CUDA_VERSION:-10.0}
      - CUDNN_VERSION=${CUDNN_VERSION:-7.5}
      - NCCL_VERSION=${NCCL_VERSION:-2.4}
      - GCC_CPU_ARCH=${GCC_CPU_ARCH:-core2}
```

I'm not familiar enough with `bazel` to know if you can just drop this underlying bazel setup inside an M1 Mac and have it "just work". My hunch is that this would end up being a lot more complicated, because we wouldn't have the legwork done by that pre-configured Ubuntu Docker repo.

### Solution 2: Upgrading TF

> Perhaps, even, tensorflow(/numpy/pytorch) has already got their act together in the latest release, and the onus is on us to unpin from the old version?

For this suggestion, if we wanted to upgrade to an M1-compatible version of Tensorflow, I presume that it would be a bleeding edge (i.e. late 2.X) version of TF. 

(Digging a little deeper, I've found https://github.com/tensorflow/tensorflow/issues/46178, which seems to suggest Tensorflow 2.4 is the first version that compiles without error on M1 Macs).

We're using 1.X right now, and I'm a little worried the upgrade path from 1.X to late 2.X won't be easy. I had [started trying out TF 2.X versions](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3367#issuecomment-907357925) but hit a snag, and put that task aside. At the very least, it will require some experimenting. 

Still, I wager this would be faster than waiting for new `sct_deepseg` methods to be developed. So, I'd definitely be on board with trying to figure this out as a... medium-term solution? In other words:

* **Short-term:** Revert back to Python 3.6/Tensorflow 1.5 (which would let us release new versions without breaking support)
* **Medium-term:** Try to upgrade to Tensorflow >2.4 (preferably >2.5, because it would [unlock Python 3.9](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3367)).
* **Long-term:** Drop Tensorflow entirely.

How does that sound?

**Comment by joshuacwnewton** [comment]
> I'm not familiar enough with `bazel` to know if you can just drop this underlying bazel setup inside an M1 Mac and have it "just work". My hunch is that this would end up being a lot more complicated, because we wouldn't have the legwork done by that pre-configured Ubuntu Docker repo.

Actually, my hesitation may have been unfounded. Look at this: https://github.com/tensorflow/tensorflow/issues/46044#issuecomment-753549970:

> Not sure if this is obvious to others but it wasn't completely obvious to me: if you install Bazel (the regular x86-64 version for mac) and use it through Rosetta 2 to compile Tensorflow from source, then it will compile without the (unsupported by Rosetta 2) AVX instruction set. I'm currently able to run both Tensorflow 1.15.0 and Tensorflow 2.5.0 using this method. Hope this is of any help :)

**Comment by joshuacwnewton** [comment]
In fact, someone has [already linked](https://github.com/tensorflow/tensorflow/issues/46044#issuecomment-797151218) a Tensorflow 1.15 wheel compiled for M1 macs, which we might be able to use in much the same way we're using the Linux AVX-less TF wheel:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/88a72e07ad7593c0f26b9e04a962201b4c5a5533/install_sct#L647-L650

**Comment by joshuacwnewton** [comment]
I've tried adding this wheel to `install_sct` in the [`jn/3566-install-m1-compiled-tensorflow`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/tree/jn/3566-install-m1-compiled-tensorflow) branch (specifically, 88bbbdc704511b3a28f177bb2f33e461d90e180d), and asked Mathieu to try installing from the branch [here](https://neuropoly.slack.com/archives/CAW9X21D5/p1636222731073700).

I do wish there was an easier was to test M1 Macs, but alas. (https://github.com/actions/virtual-environments/issues/2187, https://github.com/actions/runner/issues/805).

**Comment by kousu** [comment]
Wonderful, thanks @joshuacwnewton, that seems like a much better solution than downgrading.

**Comment by valosekj** [comment]
> I've tried adding this wheel to `install_sct` in the [`jn/3566-install-m1-compiled-tensorflow`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/tree/jn/3566-install-m1-compiled-tensorflow) branch (specifically, [88bbbdc](https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/88bbbdc704511b3a28f177bb2f33e461d90e180d)), and asked Mathieu to try installing from the branch [here](https://neuropoly.slack.com/archives/CAW9X21D5/p1636222731073700).
> 
> I do wish there was an easier was to test M1 Macs, but alas. ([actions/virtual-environments#2187](https://github.com/actions/virtual-environments/issues/2187), [actions/runner#805](https://github.com/actions/runner/issues/805))

I was installing SCT on my new M1 MacBook (MacOS 11.6) and installation failed on Keras "Illegal instruction".

I have tried to install SCT from both `master` and [`jn/3566-install-m1-compiled-tensorflow`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/tree/jn/3566-install-m1-compiled-tensorflow) branches.

<details> <summary>output for master branch installation</summary>

```

Validate installation...


--
Spinal Cord Toolbox (git-master-c9e4e80128f31755c6c9348e13de90b2364d5aee)

sct_check_dependencies 
--

SCT info:
- version: git-master-c9e4e80128f31755c6c9348e13de90b2364d5aee
- path: /usr/local/lib/sct_latest
OS: osx (Darwin-20.6.0-x86_64-i386-64bit)
CPU cores: Available: 8, Used by ITK functions: 8
RAM: Total: 16384MB, Used: 1593MB, Available: 1282MB
Check Python executable.............................[OK]
  Using bundled python 3.7.11 (default, Jul 27 2021, 07:03:16) 
[Clang 10.0.0 ] at /usr/local/lib/sct_latest/python/envs/venv_sct/bin/python
Check if data are installed.........................[OK]
Failed to extract font properties from /System/Library/Fonts/LastResort.otf: tuple indices must be integers or slices, not str
Failed to extract font properties from /System/Library/Fonts/Apple Color Emoji.ttc: In FT2Font: Could not set the fontsize (error code 0x17)
Failed to extract font properties from /System/Library/Fonts/Supplemental/NISC18030.ttf: In FT2Font: Could not set the fontsize (error code 0x17)
generated new fontManager
Check if colored is installed.......................[OK] (1.4.3)
Check if dipy is installed..........................[OK] (1.4.1)
Check if h5py is installed..........................[OK] (2.10.0)
Check if Keras (2.3.1) is installed................../install_sct: line 759: 21611 Illegal instruction: 4  sct_check_dependencies

Installation validation Failed!
```
</details>

<details> <summary>output for  jn/3566-install-m1-compiled-tensorflow branch installation</summary>

```

Validate installation...


--
Spinal Cord Toolbox (git-jn/3566-install-m1-compiled-tensorflow-56157af58b05e414f07eda63145cddafe15a4650)

sct_check_dependencies 
--

SCT info:
- version: git-jn/3566-install-m1-compiled-tensorflow-56157af58b05e414f07eda63145cddafe15a4650
- path: /usr/local/lib/sct_latest
OS: osx (Darwin-20.6.0-x86_64-i386-64bit)
CPU cores: Available: 8, Used by ITK functions: 8
RAM: Total: 16384MB, Used: 1717MB, Available: 1307MB
Check Python executable.............................[OK]
  Using bundled python 3.7.11 (default, Jul 27 2021, 07:03:16) 
[Clang 10.0.0 ] at /usr/local/lib/sct_latest/python/envs/venv_sct/bin/python
Check if data are installed.........................[OK]
Check if colored is installed.......................[OK] (1.4.3)
Check if dipy is installed..........................[OK] (1.4.1)
Check if h5py is installed..........................[OK] (2.10.0)
Check if Keras (2.3.1) is installed................../install_sct: line 745: 23122 Illegal instruction: 4  sct_check_dependencies

Installation validation Failed!
```
</details>

**Comment by kousu** [comment]
@jcohenadad can we budget to get a M1 mac in the lab? And while we're at it, a Windows 10 laptop we can wipe and erase as needed for debugging these issues?

**Comment by joshuacwnewton** [comment]
@valosekj My apologies. The `jn/3566-install-m1-compiled-tensorflow` isn't quite ready yet. 

The wheel has been tested and we've confirmed that it fixes the issue, but there's separate bash-related issue in #3569 that I have yet to sit down and figure out.

**Comment by valosekj** [comment]
No problem. I have found https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3569#discussion_r745693563 and it helped me:

<details><summary> sct_check_dependencies after manual tensorflow installation</summary>

```

sct_check_dependencies 
--

SCT info:
- version: git-jn/3566-install-m1-compiled-tensorflow-56157af58b05e414f07eda63145cddafe15a4650
- path: /usr/local/lib/sct_latest
OS: osx (Darwin-20.6.0-x86_64-i386-64bit)
CPU cores: Available: 8, Used by ITK functions: 8
RAM: Total: 16384MB, Used: 1714MB, Available: 1306MB
Check Python executable.............................[OK]
  Using bundled python 3.7.11 (default, Jul 27 2021, 07:03:16) 
[Clang 10.0.0 ] at /usr/local/lib/sct_latest/python/envs/venv_sct/bin/python
Check if data are installed.........................[OK]
Check if colored is installed.......................[OK] (1.4.3)
Check if dipy is installed..........................[OK] (1.4.1)
Check if h5py is installed..........................[OK] (2.10.0)
Check if Keras (2.3.1) is installed.................[OK] (2.3.1)
Check if ivadomed is installed......................[OK] (2.7.4)
Check if matplotlib is installed....................[OK] (3.5.0)
Check if nibabel is installed.......................[OK] (3.2.1)
Check if numpy is installed.........................[OK] (1.18.5)
Check if onnxruntime (1.4.0) is installed...........[OK] (1.4.0)
Check if pandas is installed........................[OK] (1.3.4)
Check if psutil is installed........................[OK] (5.8.0)
Check if pyqt5 (5.11.3) is installed................[OK] (5.11.3)
Check if pytest is installed........................[OK] (6.2.5)
Check if pytest-cov is installed....................[OK] (3.0.0)
Check if raven is installed.........................[OK]
Check if requests is installed......................[OK] (2.26.0)
Check if requirements-parser is installed...........[OK] (0.2.0)
Check if scipy is installed.........................[OK] (1.7.2)
Check if scikit-image is installed..................[OK] (0.18.3)
Check if scikit-learn is installed..................[OK] (1.0.1)
Check if tensorflow is installed....................[OK] (1.15.0)
Check if torch (1.5.0) is installed.................[OK] (1.5.0)
Check if torchvision (0.6.0) is installed...........[OK] (0.6.0)
Check if xlwt is installed..........................[OK] (1.3.0)
Check if tqdm is installed..........................[OK] (4.62.3)
Check if transforms3d is installed..................[OK] (0.3.1)
Check if urllib3 is installed.......................[OK] (1.26.7)
Check if pytest_console_scripts is installed........[OK]
Check if wquantiles is installed....................[OK] (0.4)
Check if spinalcordtoolbox is installed.............[OK]
Check ANTs compatibility with OS ...................[OK]
Check PropSeg compatibility with OS ................[OK]
Check if figure can be opened with matplotlib.......[OK] (Using GUI backend: 'MacOSX')
Check if figure can be opened with PyQt.............[OK]
```
</details>
